internal/vendors/equinix/equinix.go

```diff
@@ -0,0 +1,204 @@
+package equinix
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/v53/github"
+	"github.com/metal-toolbox/firmware-syncer/internal/config"
+	"github.com/metal-toolbox/firmware-syncer/internal/inventory"
+	"github.com/metal-toolbox/firmware-syncer/internal/vendors"
+	"golang.org/x/oauth2"
+
+	"github.com/pkg/errors"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/fs/operations"
+	"github.com/sirupsen/logrus"
+	serverservice "go.hollow.sh/serverservice/pkg/api/v1"
+)
+
+// Equinix implements the Vendor interface methods to retrieve Equinix OpenBMC firmware files
```

Comment:
I'd like to suggest a difference approach here - create a generic `Github` provider or downloader that is invoked if the URL begins with `https://github.com / http://www.github.com`. This would then allow the tool to be flexible in use, and it pushes the customisation of how to download from source into a destination into the configuration.

Reply:
I like this suggestion and trying to make things generic and re-usable but I think the right time to do that generalisation is when we need to implement a second provider that also uses github releases. That way we have 2 concrete implementations that will have to share the re-usable code. IMO, trying to anticipate all possible use cases in a generic abstraction before having the concrete implementations leads to big re-writes of the abstraction and all of their users.

One example that comes to mind is the Downloader abstraction that used to exist in the FUP codebase. It did the job for Dells repositories and individual firmware files but I found it difficult to expand the scope of the Downloader abstraction to cover providers that came after.

I'd be happy to discuss this further and find a good way to redesign this tool with more generic mechanisms that could be re-used across all of the providers. Currently we have 6 different providers to support (`dell`, `supermicro`, `asrockrack`, `mellanox`, `intel` and `openbmc`) and each one of them have their own quirks and ways of copying fw files from the source to the destination. My solution for that (and I admit it's not a good one) was to keep the specifics for each provider in their Sync() method and re-use functions in the downloader package.

Reply:
> I think the right time to do that generalisation is when we need to implement a second provider that also uses github releases. That way we have 2 concrete implementations that will have to share the re-usable code

Ack, this sounds like a decent approach and we can keep going with it.

> It did the job for Dells repositories and individual firmware files but I found it difficult to expand the scope of the Downloader abstraction to cover providers that came after.

Agreed, that first effort tried too hard to abstract downloading files from different vendors and did not necessarily succeed.

> My solution for that (and I admit it's not a good one) was to keep the specifics for each provider in their Sync() method and re-use functions in the downloader package.

Yep, it seems like each vendor is going to have some quirks to be handled and so we need that vendor abstraction there. If we are to focus on anything that would be generic, it would be the downloader package.

The tooling works as expected for our use case, so we're not in any rush, we can take our time with redesigning and changing its implementation - but it would be worth having a direction. And I would suggest we try to not tie anything specific to Equinix in there, that way it makes it easier to keep using this project internally while accepting external PRs as an opensource project.

I looked at the current sync source and destinations and I'll leave these here for future reference,
since I think we may be able de-dup some vendors with no specific differences, 

ASRR - `s3` to `s3` 
Dell, SMC, Mellanox, Intel `http` to `s3`
OpenBMC - `Github (http)` to `s3`

Reply:
To add to the list, the downloader functionality also have to deal with unauthenticated http endpoint (dell, smc, mellanox, intel) and authenticated ones (github). 

Also it would need to handle download of the firmware binary (dell, openbmc, asrockrack) and an archive (zip, maybe even a tar.gz) that contains the firmware binary (smc, mellanox, intel)